### PR TITLE
Add-response-model

### PR DESCRIPTION
**Pull request by dreinon**

### CODE REVIEW

File: postgrest_py/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    def __str__(self):
+        error_text = f"Error {self.code}:" if self.code else ""
+        message_text = f"\nMessage: {self.message}" if self.message else ""
+        hint_text = f"\nHint: {self.hint}" if self.hint else ""
+        details_text = f"\nDetails: {self.details}" if self.details else ""
+        complete_error_text = f"{error_text}{message_text}{hint_text}{details_text}"
+        return complete_error_text or "Empty error"
+
+    def json(self):
```
Comment (anand2312) [inline postgrest_py/exceptions.py:28]:
```suggestion
    def json(self) -> dict[str, str]:
```

File: postgrest_py/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    def __init__(self, error: dict[str, str]) -> None:
+        self._raw_error = error
+        self.message = error["message"]
+        self.code = error["code"]
+        self.hint = error["hint"]
+        self.details = error["details"]
+        super().__init__(str(self))
+
+    def __str__(self):
```
Comment (anand2312) [inline postgrest_py/exceptions.py:20]:
```suggestion
    def __str__(self) -> str:
```

Comment (anand2312) [reply postgrest_py/exceptions.py:20]: Maybe we could implement `__repr__` instead?

File: postgrest_py/base_request_builder.py [hunk 3]
```diff
@@ -90,21 +86,54 @@ def pre_delete(
...
+            "content-range"
+        )
+        if is_count_in_prefer_header and content_range_header:
+            return cls.get_count_from_content_range_header(content_range_header)
+        return None
+
+    @classmethod
+    def from_http_request_response(
+        cls: Type[APIResponse], request_response: RequestResponse
+    ):
```
Comment (anand2312) [inline postgrest_py/base_request_builder.py:133]:
```suggestion
    ) -> APIResponse:
```

**Review by anand2312 — COMMENTED** [review]
A quick first review, a few return annotations are missing

File: postgrest_py/_async/request_builder.py [hunk 4]
```diff
@@ -36,10 +38,13 @@ async def execute(self) -> Tuple[Any, Optional[int]]:
             self.path,
             json=self.json,
         )
-        return process_response(self.session, r)
+        try:
+            return APIResponse.from_http_request_response(r)
```
Comment (leynier) [inline postgrest_py/_async/request_builder.py:42]: The returning type hint of the `execute` method should be updated to `APIResponse`

File: postgrest_py/_async/request_builder.py [hunk 5]
```diff
@@ -2,7 +2,10 @@
 
 from typing import Any, Optional, Tuple
 
+from postgrest_py.exceptions import APIError
```
Comment (leynier) [inline postgrest_py/_async/request_builder.py:5]: Use relative imports instead of absolute imports

File: postgrest_py/_async/request_builder.py [hunk 6]
```diff
@@ -36,10 +38,13 @@ async def execute(self) -> Tuple[Any, Optional[int]]:
...
         )
-        return process_response(self.session, r)
+        try:
+            return APIResponse.from_http_request_response(r)
+        except ValueError as e:
+            raise APIError(r.json()) from e
 
 
-class AsyncFilterRequestBuilder(BaseFilterRequestBuilder, AsyncQueryRequestBuilder):
+class AsyncFilterRequestBuilder(BaseFilterRequestBuilder, AsyncQueryRequestBuilder):  # type: ignore
```
Comment (leynier) [inline postgrest_py/_async/request_builder.py:47]: Why is necessary to ignore type checking?

File: postgrest_py/base_request_builder.py [hunk 7]
```diff
@@ -90,21 +86,54 @@ def pre_delete(
...
+        prefer_header: Optional[str] = request_response.request.headers.get("prefer")
+        if not prefer_header:
+            return None
+        is_count_in_prefer_header = cls.is_count_in_prefer_header(prefer_header)
+        content_range_header: Optional[str] = request_response.headers.get(
+            "content-range"
+        )
+        if is_count_in_prefer_header and content_range_header:
+            return cls.get_count_from_content_range_header(content_range_header)
+        return None
```
Comment (leynier) [inline postgrest_py/base_request_builder.py:128]: Unnecessary return of None, by default all execution of functions without explicit return instruction return None.

File: postgrest_py/exceptions.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    hint: str
+    details: str
+
+    def __init__(self, error: Dict[str, str]) -> None:
+        self._raw_error = error
+        self.message = error["message"]
+        self.code = error["code"]
+        self.hint = error["hint"]
+        self.details = error["details"]
+        super().__init__(str(self))
```
Comment (leynier) [inline postgrest_py/exceptions.py:21]: Use `Exception.__init__(self, str(self))` to avoid possible future problems with multiple inheritances

**Review by leynier — CHANGES_REQUESTED** [review]

File: postgrest_py/_async/request_builder.py (hunk 6 above)
Comment (dreinon) [reply postgrest_py/_async/request_builder.py:47]: In this [commit](https://github.com/supabase-community/postgrest-py/commit/790811f00764b3fb1734bfdb25964972ed93a2ca) I tagged the mypy issue.

File: postgrest_py/base_request_builder.py (hunk 7 above)
Comment (dreinon) [reply postgrest_py/base_request_builder.py:128]: My reference was the following: [https://stackoverflow.com/questions/15300550/return-return-none-and-no-return-at-all](https://stackoverflow.com/questions/15300550/return-return-none-and-no-return-at-all)

File: postgrest_py/_async/request_builder.py (hunk 6 above)
Comment (leynier) [reply postgrest_py/_async/request_builder.py:47]: Please, add the reference like a comment in the code.

Comment (dreinon) [reply postgrest_py/_async/request_builder.py:47]: Just by writing the url next to type: ignore? @leynier

**Review by anand2312 — COMMENTED** [review]
This looks good to me 🚀 
Nitpick: What do you think about making the classmethods/staticmethods on the APIResponse private (adding a leading `_`)

**Review by leynier — APPROVED** [review]

**Comment by dreinon** [comment]
> This looks good to me 🚀 
> Nitpick: What do you think about making the classmethods/staticmethods on the APIResponse private (adding a leading `_`)

@anand2312 Done!

**Pull request merged**